cmd/kola/updatepayload.go

```diff
@@ -0,0 +1,239 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/coreos/mantle/kola"
+	"github.com/coreos/mantle/network/omaha"
+	"github.com/coreos/mantle/platform"
+	"github.com/coreos/mantle/sdk"
+	sdkomaha "github.com/coreos/mantle/sdk/omaha"
+	"github.com/coreos/mantle/util"
+
+	"github.com/coreos/mantle/Godeps/_workspace/src/github.com/spf13/cobra"
+)
+
+func init() {
+	root.AddCommand(cmdUpdatePayload)
+}
+
+var cmdUpdatePayload = &cobra.Command{
+	Run:   runUpdatePayload,
+	Use:   "updatepayload",
+	Short: "test serving a update_engine payload",
+	Long: `
+Boot a CoreOS instance and serve an update payload to its update_engine.
+
+This command must run inside of the SDK as root, e.g.
+
+sudo kola updatepayload
+`,
+}
+
+var userdata = `#cloud-config
+
+coreos:
+  update:
+    server: "http://{{.Server}}/v1/update/"
+    # we disable reboot so we have explicit control
+    reboot-strategy: "off"
+`
+
+func runUpdatePayload(cmd *cobra.Command, args []string) {
+	if len(args) != 0 {
+		plog.Fatal("No args accepted")
+	}
+
+	plog.Info("Generating update payload")
+
+	// check for update file, generate if it doesn't exist
+	version := "latest"
+	dir := sdk.BuildImageDir(version)
+	payload := "coreos_production_update.gz"
+
+	_, err := os.Stat(filepath.Join(dir, payload))
+	if err != nil {
+		err = sdkomaha.GenerateFullUpdate("latest", true)
+		if err != nil {
+			plog.Fatalf("Building full update failed: %v", err)
+		}
+	}
+
+	plog.Info("Bringing up test harness cluster")
+
+	cluster, err := platform.NewQemuCluster(kola.QEMUOptions)
+	qc := cluster.(*platform.QEMUCluster)
+	if err != nil {
+		plog.Fatalf("Cluster failed: %v", err)
+	}
+	defer cluster.Destroy()
+
+	svc := &updateServer{
+		updatePath: dir,
+		payload:    payload,
+	}
+
+	qc.OmahaServer.Updater = svc
+
+	// tell omaha server to handle file requests for the images dir
+	qc.OmahaServer.Mux.Handle(dir+"/", http.StripPrefix(dir+"/", http.FileServer(http.Dir(dir))))
+
+	_, port, err := net.SplitHostPort(qc.OmahaServer.Addr().String())
+	if err != nil {
+		plog.Errorf("SplitHostPort failed: %v", err)
+		return
+	}
+
+	tmplVals := map[string]string{
+		"Server": fmt.Sprintf("10.0.0.1:%s", port),
+	}
+
+	tmpl := template.Must(template.New("userdata").Parse(userdata))
+	buf := new(bytes.Buffer)
+
+	err = tmpl.Execute(buf, tmplVals)
+	if err != nil {
+		plog.Errorf("Template execution failed: %v", err)
+		return
+	}
+
+	plog.Infof("Spawning test machine")
+
+	m, err := cluster.NewMachine(buf.String())
+	if err != nil {
+		plog.Errorf("Machine failed: %v", err)
+		return
+	}
+
+	plog.Info("Checking for boot from USR-A partition")
+
+	/* check that we are on USR-A. */
+	out, err := m.SSH("findmnt -no PARTUUID /usr")
```

Comment:
This isn't going to work when we enable verity since there is a dm layer in between. Maybe just a simple minded grep $USRAUUID /proc/cmdline will suffice.